core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -650,14 +650,24 @@ private IndexResult innerIndex(Index index) throws IOException {
                 }
                 updatedVersion = index.versionType().updateVersion(currentVersion, expectedVersion);
                 index.parsedDoc().version().setLongValue(updatedVersion);
+
+                // Update the document's sequence number and primary term, the
+                // sequence number here is derived here from either the sequence
+                // number service if this is on the primary, or the existing
+                // document's sequence number if this is on the replica. The
+                // primary term here has already been set, see
+                // IndexShard.prepareIndex where the Engine.Index operation is
+                // created
+                index.parsedDoc().updateSeqID(seqNo, index.primaryTerm());
```

Comment:
another drive by question - should we do it here, or separate this into two, more explicit, flows - one we when we create an Index operation on a replica (where we set the seq no number based on incoming request) and one here when we operate as a primary? If you guys feel the current version is more intuitive, I'm good but I wanted to get confirmation this was considered.